File: alert_handler_test.go
package main

import "testing"
import "github.com/stretchr/testify/assert"

func Test_UnknownDevices(t *testing.T) {
	// Given
	repository := deviceRepositoryTestAdapter{make(map[string]Device)}
	handler := alertHandler{repository}
	// When
	handler.onNetworkInterfaceAlert("device", "other device")
	assert.Equal(t, 1, repository.findDevice("device").alertsCount)
	assert.Equal(t, 1, repository.findDevice("other device").alertsCount)
}

func Test_KnownDevices(t *testing.T) {
	// Given
	repository := deviceRepositoryTestAdapter{make(map[string]Device)}
	repository.devices["device"] = Device{alertsCount:22, name:"device"}
	handler := alertHandler{repository}
	// When
	handler.onNetworkInterfaceAlert("device", "other device")
	assert.Equal(t, 23, repository.findDevice("device").alertsCount)
}

func Test_UpdateNeighbour(t *testing.T) {
	// Given
	repository := deviceRepositoryTestAdapter{make(map[string]Device)}
	repository.devices["device"] = Device{alertsCount:22, name:"device"}
	handler := alertHandler{repository}
	// When
	handler.onNetworkInterfaceAlert("device", "other device")
	assert.Equal(t, "other device", repository.findDevice("device").mostFailingNeighbour.name)
	assert.Equal(t, 1, repository.findDevice("device").mostFailingNeighbour.alertsCount)
	assert.Equal(t, "device", repository.findDevice("other device").mostFailingNeighbour.name)
	assert.Equal(t, 23, repository.findDevice("other device").mostFailingNeighbour.alertsCount)
}

func Test_NotUpdateUpdateNeighbour(t *testing.T) {
	// Given
	repository := deviceRepositoryTestAdapter{make(map[string]Device)}
	thirdDevice := Device{alertsCount: 10, name: "third device"}
	repository.devices["third device"] = thirdDevice
	repository.devices["device"] = Device{alertsCount:22, name:"device", mostFailingNeighbour:&thirdDevice}
	handler := alertHandler{repository}
	// When
	handler.onNetworkInterfaceAlert("device", "other device")
	assert.Equal(t, "third device", repository.findDevice("device").mostFailingNeighbour.name)
	assert.Equal(t, 10, repository.findDevice("device").mostFailingNeighbour.alertsCount)
}

type deviceRepositoryTestAdapter struct {
	devices map[string]Device
}

func (r deviceRepositoryTestAdapter) findDevice(deviceName string) Device {
	return r.devices[deviceName]
}

func (r deviceRepositoryTestAdapter) updateAlertsCount(deviceName string, alertsCount int) {
	device := r.devices[deviceName]
	device.alertsCount = alertsCount
	r.devices[deviceName] = device
}

func (r deviceRepositoryTestAdapter) updateMostFailingNeighbour(deviceName string, neighbour Device) {
	device := r.devices[deviceName]
	device.mostFailingNeighbour = &neighbour
	r.devices[deviceName] = device
}

File: src/test/java/com/orange/craftsmanship/alarming/domain/AlertHandlerTest.java
package com.orange.craftsmanship.alarming.domain;

import org.junit.Before;
import org.junit.Test;

import static java.util.Optional.of;
import static org.mockito.Mockito.*;

public class AlertHandlerTest {

    private DeviceRepositoryPort alarmRepository = mock(DeviceRepositoryPort.class);
    private AlertHandler handler;

    @Before
    public void setUp() {
        handler = new AlertHandler(alarmRepository);
    }

    @Test
    public void shall_update_alerts_count_for_unknown_devices() {
        // When
        handler.onNetworkInterfaceAlert("first device", "second device");
        // Then
        verify(alarmRepository).updateAlertsCount("first device", 1);
        verify(alarmRepository).updateAlertsCount("second device", 1);
    }

    @Test
    public void shall_update_alerts_count_for_known_devices() {
        // Given
        when(alarmRepository.findDevice("device")).thenReturn(of(createDevice("device", 3)));
        when(alarmRepository.findDevice("other device")).thenReturn(of(createDevice("other device", 10)));
        // When
        handler.onNetworkInterfaceAlert("device", "other device");
        // Then
        verify(alarmRepository).updateAlertsCount("device", 4);
        verify(alarmRepository).updateAlertsCount("other device", 11);
    }

    @Test
    public void shall_update_most_failing_neighbour() {
        // Given
        Device device = createDevice("device", 3);
        when(alarmRepository.findDevice("device")).thenReturn(of(device));
        Device otherDevice = createDevice("other device", 10);
        when(alarmRepository.findDevice("other device")).thenReturn(of(otherDevice));
        // When
        handler.onNetworkInterfaceAlert("device", "other device");
        // Then
        verify(alarmRepository).updateMostFailingNeighbour("device", otherDevice);
        verify(alarmRepository).updateMostFailingNeighbour("other device", device);
    }

    @Test
    public void shall_not_update_most_failing_neighbour() {
        // Given
        Device device = createDevice("device", 3);
        device.mostFailingNeighbour = of(createDevice("third device", 20));
        when(alarmRepository.findDevice("device")).thenReturn(of(device));
        when(alarmRepository.findDevice("other device")).thenReturn(of(createDevice("other device", 10)));
        // When
        handler.onNetworkInterfaceAlert("device", "other device");
        // Then
        verify(alarmRepository, times(0)).updateMostFailingNeighbour(eq("device"), any());
    }

    private Device createDevice(String name, int alertsCount) {
        Device device = new Device(name);
        device.alertsCount = alertsCount;
        return device;
    }
}
